### PR TITLE
Enable faster CUDA kernels for the causal product

### PR DESCRIPTION
**Pull request by jdemouth**
That pull-request introduces new CUDA kernels for the causal product. The new kernels move the loop(s) iterating over the sequence length entirely to the GPU and do register blocking to keep K*V outer products. The performance of those kernels is higher than existing kernels.

### CODE REVIEW

**Comment by angeloskath** [comment]
This PR is merged in 68d26c0 and it improves our causal_dot_product kernels massively.

The following simple benchmark script includes the normalization and everything (but not the linear projections) and shows a clear >4x improvement on my RTX 2060S.

```python
import torch

from fast_transformers.attention import CausalLinearAttention
from fast_transformers.masking import TriangularCausalMask, FullMask


def measure_time(L, E, M, H):
    queries = torch.randn(10, L, H, E).cuda()
    keys = torch.randn(10, L, H, E).cuda()
    values = torch.randn(10, L, H, M).cuda()
    attn_mask = TriangularCausalMask(L, device="cuda")
    length_mask = FullMask(10, L, device="cuda")
    attn = CausalLinearAttention(64)

    # warmup
    attn(queries, keys, values, attn_mask, length_mask, length_mask)

    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)

    start.record()
    for i in range(10):
        attn(queries, keys, values, attn_mask, length_mask, length_mask)
    end.record()
    torch.cuda.synchronize()

    return start.elapsed_time(end)


if __name__ == "__main__":
    for l in [128, 256, 512, 1024, 2048, 4096, 8192, 16384]:
        print(l, measure_time(l, 64, 64, 8))
```

Output with the old kernel:
```
128 17.475168228149414
256 34.6049919128418
512 68.11033630371094
1024 117.60233306884766
2048 211.506591796875
4096 424.8102111816406
8192 851.4991455078125
16384 1702.237060546875
```

and with the new:

```
128 3.5999040603637695
256 6.929247856140137
512 13.47379207611084
1024 26.533472061157227
2048 52.61721420288086
4096 104.73661041259766
8192 186.431640625
16384 391.29534912109375
```
